crypto/asn1/tasn_dec.c

```diff
@@ -966,6 +990,84 @@ static int asn1_find_end(const unsigned char **in, long len, char inf) {
   return 1;
 }
 
+// This function collects the asn1 data from a constructed string type into
+// a buffer. The values of 'in' and 'len' should refer to the contents of the
+// constructed type and 'inf' should be set if it is indefinite length.
+
+// This determines how many levels of recursion are permitted in ASN1 string
+// types. If it is not limited stack overflows can occur. If set to zero no
+// recursion is allowed at all. Although zero should be adequate examples
+// exist that require a value of 1. So 5 should be more than enough.
+
+#define ASN1_MAX_STRING_NEST 5
+
+static int asn1_collect(BUF_MEM *buf, const unsigned char **in, long len,
+                        char inf, int tag, int aclass, int depth) {
+  const unsigned char *p, *q;
+  long plen;
+  char cst, ininf;
+  p = *in;
+  inf &= 1;
+  // If no buffer and not indefinite length constructed just pass over the
+  // encoded data
+  if (!buf && !inf) {
+    *in += len;
+    return 1;
+  }
+  while (len > 0) {
+    q = p;
+    // Check for EOC.
+    if (asn1_check_eoc(&p, len)) {
+      // EOC is illegal outside indefinite length constructed form
+      if (!inf) {
+        OPENSSL_PUT_ERROR(ASN1, ASN1_R_UNEXPECTED_EOC);
+        return 0;
+      }
+      inf = 0;
+      break;
+    }
+
+    if (!asn1_check_tlen(&plen, NULL, NULL, &ininf, &cst, &p,
+                         len, tag, aclass, 0)) {
+      OPENSSL_PUT_ERROR(ASN1, ASN1_R_NESTED_ASN1_ERROR);
+      return 0;
+    }
+
+    // If indefinite length constructed, update max length.
+    if (cst) {
+      if (depth >= ASN1_MAX_STRING_NEST) {
+        OPENSSL_PUT_ERROR(ASN1, ASN1_R_NESTED_ASN1_STRING);
+        return 0;
+      }
+      if (!asn1_collect(buf, &p, plen, ininf, tag, aclass, depth + 1)) {
+        return 0;
+      }
+    } else if (plen && !collect_data(buf, &p, plen)) {
+      return 0;
+    }
+    len -= p - q;
+  }
+  if (inf) {
+    OPENSSL_PUT_ERROR(ASN1, ASN1_R_MISSING_EOC);
+    return 0;
+  }
+  *in = p;
+  return 1;
+}
+
+static int collect_data(BUF_MEM *buf, const unsigned char **p, long plen) {
+  int len;
+  if (buf) {
```

Comment:
It seems like `collect_data` is never called with `buf == NULL`, should this return `0` instead?

Reply:
Great question, I did some digging around. But judging from the comment in `asn1_collect` which mentions "If no buffer and not indefinite length constructed just pass over the encoded data", it does seem like there's the possibility of `buf` being `NULL` and there being operations that need to be done under that circumstance.

I'll leave it as is for the time being, OpenSSL's ASN1 implementation is also like this today.